# Exploratory Data Analysis Case Study - 
##### Conducted by Nirbhay Tandon & Naveen Sharma

#### Case study Brief: 

## 1.Import libraries and set required parameters

In [1]:
#import all the libraries and modules
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import re
from scipy import stats 

# Supress Warnings
#Enable autocomplete in Jupyter Notebook.
%config IPCompleter.greedy=True

import warnings
warnings.filterwarnings('ignore')
import os

In [2]:
## Set the max display columns to None so that pandas doesn't sandwich the output 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

df=pd.read_csv("./application_data.csv")

df.head()

SK_ID_CURR  TARGET NAME_CONTRACT_TYPE CODE_GENDER FLAG_OWN_CAR  \
0      100002       1         Cash loans           M            N   
1      100003       0         Cash loans           F            N   
2      100004       0    Revolving loans           M            Y   
3      100006       0         Cash loans           F            N   
4      100007       0         Cash loans           M            N   

  FLAG_OWN_REALTY  CNT_CHILDREN  AMT_INCOME_TOTAL  AMT_CREDIT  AMT_ANNUITY  \
0               Y             0          202500.0    406597.5      24700.5   
1               N             0          270000.0   1293502.5      35698.5   
2               Y             0           67500.0    135000.0       6750.0   
3               Y             0          135000.0    312682.5      29686.5   
4               Y             0          121500.0    513000.0      21865.5   

   AMT_GOODS_PRICE NAME_TYPE_SUITE NAME_INCOME_TYPE  \
0         351000.0   Unaccompanied          Working   
1        1129500.0          Family    State servant   
2         135000.0   Unaccompanied          Working   
3         297000.0   Unaccompanied          Working   
4         513000.0   Unaccompanied          Working   

             NAME_EDUCATION_TYPE    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  \
0  Secondary / secondary special  Single / not married  House / apartment   
1               Higher education               Married  House / apartment   
2  Secondary / secondary special  Single / not married  House / apartment   
3  Secondary / secondary special        Civil marriage  House / apartment   
4  Secondary / secondary special  Single / not married  House / apartment   

   REGION_POPULATION_RELATIVE  DAYS_BIRTH  DAYS_EMPLOYED  DAYS_REGISTRATION  \
0                    0.018801       -9461           -637            -3648.0   
1                    0.003541      -16765          -1188            -1186.0   
2                    0.010032      -19046           -225            -4260.0   
3                    0.008019      -19005          -3039            -9833.0   
4                    0.028663      -19932          -3038            -4311.0   

   DAYS_ID_PUBLISH  OWN_CAR_AGE  FLAG_MOBIL  FLAG_EMP_PHONE  FLAG_WORK_PHONE  \
0            -2120          NaN           1               1                0   
1             -291          NaN           1               1                0   
2            -2531         26.0           1               1                1   
3            -2437          NaN           1               1                0   
4            -3458          NaN           1               1                0   

   FLAG_CONT_MOBILE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  CNT_FAM_MEMBERS  \
0                 1           1           0        Laborers              1.0   
1                 1           1           0      Core staff              2.0   
2                 1           1           0        Laborers              1.0   
3                 1           0           0        Laborers              2.0   
4                 1           0           0      Core staff              1.0   

   REGION_RATING_CLIENT  REGION_RATING_CLIENT_W_CITY  \
0                     2                            2   
1                     1                            1   
2                     2                            2   
3                     2                            2   
4                     2                            2   

  WEEKDAY_APPR_PROCESS_START  HOUR_APPR_PROCESS_START  \
0                  WEDNESDAY                       10   
1                     MONDAY                       11   
2                     MONDAY                        9   
3                  WEDNESDAY                       17   
4                   THURSDAY                       11   

   REG_REGION_NOT_LIVE_REGION  REG_REGION_NOT_WORK_REGION  \
0                           0                           0   
1                           0                           0   
2                        

## 2. Data Sourcing

In [3]:
#shape of allpcation_data.csv data

df.shape

(307511, 122)

In [4]:
#take information about the data

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


In [ ]:
#get the information about the numerical data

df.describe()

In [ ]:
## print the column names for application_data.csv
df.columns


In [ ]:
## print the various datatypes of application_data.csv
df.dtypes

## 3. Data Cleaning

In [ ]:
## Check the number of null values of each column 
df.isnull().sum()

In [ ]:
## calculate percentage of null values
percentageOfNullValues = df.isnull().sum()*100/len(df)
columnNamesWithPrcntgOfNullValues = pd.DataFrame({'Column Name': df.columns, 'Percentage ':round(percentageOfNullValues,2)})
print(columnNamesWithPrcntgOfNullValues)


In [ ]:
print(df.shape)

In [ ]:
#Drop all the columns where the 
## percentage of missing values is above 60% in application_data.csv
print(df.shape)
droppedColumns = df.columns[df.isnull().mean() > 0.60]
dfAfterDroppedColumns = df.drop(droppedColumns, axis = 1)

print(dfAfterDroppedColumns.shape)
print(dfAfterDroppedColumns.columns)


### Observation: 
###### As you can see, the shape of the data has changed from (307511, 122) to (307511, 105). Which mean we have dropped 17 columns that had over 60% percent null values. The dropped columns are mentioned below.

In [ ]:
print("The columns that have been dropped are: ", droppedColumns)

In [ ]:
## print the percentage of columns with null values in the 
## new data frame after the columns have been dropped
percentageOfNullValuesAfterColDrop = dfAfterDroppedColumns.isnull().sum()*100/len(dfAfterDroppedColumns)
columnNamesWithPrcntgOfNullValuesAfterColDrop = pd.DataFrame({'Column Name': dfAfterDroppedColumns.columns, 'Percentage ':round(percentageOfNullValuesAfterColDrop,2)})
print(columnNamesWithPrcntgOfNullValuesAfterColDrop)

### Observation: 
##### As you can see above, there are still a few columns that have a above 30% of null/missing values. We can deal with those null/missing values using various methods of imputation. 


In [ ]:
## Checking the datadrame after dropping null values

In [ ]:
dfAfterDroppedColumns.head()

In [ ]:
#After dropping the columns 

dfAfterDroppedColumns.shape

In [ ]:
#cast all continous variable to numeric

df_2 = dfAfterDroppedColumns.select_dtypes(np.number)

df_2.head()

## Derived Metrics

In [ ]:
#credit income ratio

dfAfterDroppedColumns['credit_income_ratio']=dfAfterDroppedColumns['AMT_CREDIT']/dfAfterDroppedColumns['AMT_INCOME_TOTAL']
dfAfterDroppedColumns.head()

In [ ]:
#create bins for credit anount

bins=[0,500000,1000000,1500000,2000000,2500000,3000000,4000000]
range_period=['0-500000','500000-1000000','1000000-1500000','1500000-2000000','2000000-2500000','2500000-3000000','3000000-4000000']
dfAfterDroppedColumns['credit_amount_range']=pd.cut(dfAfterDroppedColumns['AMT_CREDIT'],bins,labels=range_period)


In [ ]:

bins=[0,1000000,2000000,3000000,4000000,5000000,6000000,200000000]
range_period=['0-1000000','1000000-2000000','2000000-3000000','3000000-4000000','4000000-5000000','5000000-6000000','6000000 and above']
dfAfterDroppedColumns['Income_amount_range']=pd.cut(dfAfterDroppedColumns['AMT_INCOME_TOTAL'],bins,labels=range_period)


In [ ]:
# Age calculation from the column "DAYS_BIRTH".

dfAfterDroppedColumns['Age'] = ((dfAfterDroppedColumns['DAYS_BIRTH']*-1)/365)
dfAfterDroppedColumns['Age']= dfAfterDroppedColumns['Age'].astype(int)

bins=[0,25,50,75,100]
group_names=['0-25','25-50','50-75','75-100']
dfAfterDroppedColumns['Age_binned'] = pd.cut(dfAfterDroppedColumns['Age'], bins, labels=group_names)

In [ ]:
dfAfterDroppedColumns.head()

In [ ]:
#Plot Box Plot of AMT_INCOME_TOTAL, AMT_CREDIT, AMT_ANNUITY and AMT_GOODS_PRICE columns

fig, axes = plt.subplots(nrows=2, ncols = 2, figsize=(20,12))

sns.boxplot(x="TARGET", y="AMT_INCOME_TOTAL", data= dfAfterDroppedColumns, ax=axes[0][0])
axes[0][0].set_title('AMT_INCOME_TOTAL')

sns.boxplot(x="TARGET", y="AMT_CREDIT", data= dfAfterDroppedColumns, ax=axes[0][1])
axes[0][1].set_title('AMT_CREDIT')

sns.boxplot(x="TARGET", y="AMT_ANNUITY", data= dfAfterDroppedColumns, ax=axes[1][0])
axes[1][0].set_title('AMT_ANNUITY')

sns.boxplot(x="TARGET", y="AMT_GOODS_PRICE", data= dfAfterDroppedColumns, ax=axes[1][1])
axes[1][1].set_title('AMT_GOODS_PRICE')

plt.show()


We can easily see in the box plot that there are so many outliers which has to removed for the better calculation. So, In the next part of the code we remove outliers from the function "remove_outliers" which accept dataframe and columns name (In which we want to remove outliers) as argument and return the outliers removed dataframe.

In [ ]:
#Function for removing outliers

def remove_outlier(df_2, col_name):
    q1 = df_2[col_name].quantile(0.25)
    q3 = df_2[col_name].quantile(0.75)
    iqr = q3-q1 #Interquartile range
    l  = q1-1.5*iqr
    h = q3+1.5*iqr
    df_3 = df_2.loc[(df_2[col_name] > l) & (df_2[col_name] < h)]
    return df_3

In [ ]:
#Function calling of function "remove_outliers"

cols=['AMT_INCOME_TOTAL', 'AMT_CREDIT','AMT_ANNUITY', 'AMT_GOODS_PRICE']

for i in cols:
    dfAfterDroppedColumns=remove_outlier(dfAfterDroppedColumns,i)
    
dfAfterDroppedColumns.head()

From now on we will work with this dataframe(dfAfterDroppedColumns).

In [ ]:
#Plot Box plot again after removing outliers.

fig, axes = plt.subplots(nrows=2, ncols = 2, figsize=(20,12))

sns.boxplot(x="TARGET", y="AMT_INCOME_TOTAL", data= dfAfterDroppedColumns, ax=axes[0][0])
axes[0][0].set_title('AMT_INCOME_TOTAL')

sns.boxplot(x="TARGET", y="AMT_CREDIT", data= dfAfterDroppedColumns, ax=axes[0][1])
axes[0][1].set_title('AMT_CREDIT')

sns.boxplot(x="TARGET", y="AMT_ANNUITY", data= dfAfterDroppedColumns, ax=axes[1][0])
axes[1][0].set_title('AMT_ANNUITY')

sns.boxplot(x="TARGET", y="AMT_GOODS_PRICE", data= dfAfterDroppedColumns, ax=axes[1][1])
axes[1][1].set_title('AMT_GOODS_PRICE')


plt.show()


After cleaning and null values handling in the data, our data is ready for data balancing and analysis.

We perform two type of analysis-

1- Univariate Analysis
2- Bivariate Analysis

In [ ]:
#Check the total count of target value as 0 and 1.

target_class=pd.value_counts(dfAfterDroppedColumns['TARGET'],sort=True)
target_class


In [ ]:
target_class.plot(kind='bar',rot=0)

We can easily see that this data is very much imbalance. Rows with target value 0 is only 91.62% and with 1 is only 8.38%.

For data-balancing we make two dataframes.

1- pa_df-> Having target value 1(Payment difficulties)
2- o_df -> Having target value 0(Other difficulties)

In [ ]:
(target_class[0]/dfAfterDroppedColumns.shape[0])*100

In [ ]:
(target_class[1]/dfAfterDroppedColumns.shape[0])*100

We have 242534 rows which has TARGET value as 0 and 22306 rows which has TARGET value as 1.

So, for data balancing we use below part of the code.

pa_df dataframe having TARGET value as 1.

In [ ]:
#Data balancing and then print the both balance data.

pa_df=dfAfterDroppedColumns.loc[dfAfterDroppedColumns['TARGET']==1]
zero_df=dfAfterDroppedColumns.loc[dfAfterDroppedColumns['TARGET']==0].sample(n=22306,random_state=42)
pa_df.head()

Dataframe having TARGET value as 0.

In [ ]:
zero_df.head()

In [ ]:
#Concat both the dataframes(pa_df and o_df)

normalize_df=pd.concat([pa_df,zero_df])


In [ ]:
plt.figure(figsize=(8,8))
sns.countplot('TARGET',data=normalize_df)
plt.title('Balanced')
plt.show()


In the above bar plot, we can easily see that now data is balance both value 0 and 1 has equal number of rows in the dataframe.

### Univariate Analysis:-

### Continuious variable

In [ ]:
#Plot the bar graph of total people of different ages.

dfAfterDroppedColumns['Age_binned'].value_counts().plot(kind='bar')

In [ ]:
dfAfterDroppedColumns['credit_amount_range'].value_counts().plot(kind='bar')

In [ ]:
dfAfterDroppedColumns['Income_amount_range'].value_counts().plot(kind='bar')

In [ ]:
dfAfterDroppedColumns['AMT_CREDIT'].plot.area()

### Categorical Variable

In [ ]:
#Calculate the total frquency of each INCOME_TYPE which having TARGET value as 1

freq_table=pa_df.groupby(['NAME_INCOME_TYPE']).size().reset_index(name='count').rename(columns={'NAME_INCOME_TYPE':'INCOME_TYPE'})
freq_table

In [ ]:
#Plot the graph of the count of the INCOME_TYPE which having TARGET value as 1

plt.figure(figsize=(10,8))
sns.set(style="whitegrid")
ax=sns.barplot(x='INCOME_TYPE',y='count',data=freq_table)
plt.title(['INCOME_TYPE,TARGET=1'])


In [ ]:
#Calculate the total frquency of each INCOME_TYPE which having TARGET value as 0

freq_table1=zero_df.groupby(['NAME_INCOME_TYPE']).size().reset_index(name='count').rename(columns={'NAME_INCOME_TYPE':'INCOME_TYPE'})
freq_table1

In [ ]:
#Plot the graph of the count of the INCOME_TYPE which having TARGET value as 0

plt.figure(figsize=(10,8))
sns.set(style="whitegrid")
ax=sns.barplot(x='INCOME_TYPE',y='count',data=freq_table1)
plt.title(['INCOME_TYPE,TARGET=0'])
plt.show()

In [ ]:
#Calculate the total frquency of each FAMILY_STATUS which having TARGET value as 1

freq_table2=pa_df.groupby(['NAME_FAMILY_STATUS']).size().reset_index(name='count').rename(columns={'NAME_FAMILY_STATUS':'FAMILY_STATUS'})
freq_table2

In [ ]:
#Plot the pai-graph of the count of the INCOME_TYPE which having TARGET value as 1

plt.figure(figsize=(8,8))
freq_table2['count'].groupby(freq_table2.FAMILY_STATUS).sum().plot(kind="pie")
plt.title(['INCOME_TYPE,TARGET=1'])
plt.show()

In [ ]:
#Calculate the total frquency of each FAMILY_STATUS which having TARGET value as 0

freq_table3=zero_df.groupby(['NAME_FAMILY_STATUS']).size().reset_index(name='count').rename(columns={'NAME_FAMILY_STATUS':'FAMILY_STATUS'})
freq_table3

In [ ]:
#Plot the pai-graph of the count of the INCOME_TYPE which having TARGET value as 0

plt.figure(figsize=(8,8))
freq_table3['count'].groupby(freq_table3.FAMILY_STATUS).sum().plot(kind="pie")
plt.title(['INCOME_TYPE,TARGET=0'])
plt.show()

### Bivariate Analysis:-

Perform correlation between numerical columns for finding correlation which having TARGET value as 1

In [ ]:
plt.figure(figsize=(14,10))

sns.boxplot(data=dfAfterDroppedColumns,x='NAME_CONTRACT_TYPE',y='AMT_CREDIT',hue='TARGET')
plt.title('Difficulty on credit loan of different type')
plt.show()

In [ ]:
#Perform correlation between CNT_CHILDREN, AMT_INCOME_TOTAL, AMT_CREDIT, AMT_GOODS_PRICE and REGION_POPULATION_RELATIVE
#Then make correlation matrix

columns=['CNT_CHILDREN','AMT_INCOME_TOTAL','AMT_CREDIT','AMT_GOODS_PRICE','REGION_POPULATION_RELATIVE']
corr=pa_df[columns].corr()
corr.style.background_gradient(cmap='coolwarm')

In the correlation matrix, we can identify-

Columns with High Correlation:

1.AMT_GOODS_PRICE and AMT_CREDIT

Columns with Medium Correlation:

1.REGION_POPULATION_RELATIVE and AMT_INCOME_TOTAL                                                                           
2.REGION_POPULATION_RELATIVE and AMT_GOODS_PRICE                                                         
3.REGION_POPULATION_RELATIVE and AMT_CREDIT

Columns with low correlation:

1.AMT_INCOME_TOTAL and CNT_CHILDREN

In [ ]:
#For better presentaion and understanding we plot heat map between numerical columns.

plt.figure(figsize=(15,10))
sns.heatmap(pa_df[columns].corr(), annot=True,cmap="YlGn")

Perform correlation between numerical columns for finding correlation which having TARGET value as 0

In [ ]:
#Perform correlation between CNT_CHILDREN, AMT_INCOME_TOTAL, AMT_CREDIT, AMT_GOODS_PRICE and REGION_POPULATION_RELATIVE
#Then make correlation matrix

corr=zero_df[columns].corr()
corr.style.background_gradient(cmap='coolwarm')

In the correlation matrix, we can identify-

Columns with High Correlation:

1.AMT_GOODS_PRICE and AMT_CREDIT

Columns with Medium Correlation:

1.AMT_INCOME_TOTAL and AMT_CREDIT                                                                          
2.AMT_INCOME_TOTAL and AMT_GOODS_PRICE                                                         

Columns with low correlation:

1.AMT_GOODS_PRICE and CNT_CHILDREN

In [ ]:
#For better presentaion and understanding we plot heat map between numerical columns.

plt.figure(figsize=(15,10))
sns.heatmap(zero_df[columns].corr(), annot=True,cmap="YlGn")

In [ ]:
#Box plot on the numerical columns having TARGET value as 1

plt.figure(figsize=(15,15))
plt.subplot(2,2,1)
plt.title('CHILDREN COUNT')
sns.boxplot(pa_df['CNT_CHILDREN'])


plt.subplot(2,2,2)
plt.title('AMT_INCOME_TOTAL')
sns.boxplot(pa_df['AMT_INCOME_TOTAL'])


plt.subplot(2,2,3)
plt.title('AMT_CREDIT')
sns.boxplot(pa_df['AMT_CREDIT'])

plt.subplot(2,2,4)
plt.title('AMT_GOODS_PRICE')
sns.boxplot(pa_df['AMT_GOODS_PRICE'])


plt.show()

In [ ]:
#Box plot on the numerical columns having TARGET value as 0

plt.figure(figsize=(15,15))
plt.subplot(2,2,1)
plt.title('CHILDREN COUNT')
sns.boxplot(zero_df['CNT_CHILDREN'])


plt.subplot(2,2,2)
plt.title('AMT_INCOME_TOTAL')
sns.boxplot(zero_df['AMT_INCOME_TOTAL'])


plt.subplot(2,2,3)
plt.title('AMT_CREDIT')
sns.boxplot(zero_df['AMT_CREDIT'])

plt.subplot(2,2,4)
plt.title('AMT_GOODS_PRICE')
sns.boxplot(zero_df['AMT_GOODS_PRICE'])


plt.show()

In [ ]:
#Heat map on numerical column for TARGET as 1

plt.figure(figsize=(15,12))
col1=['AMT_INCOME_TOTAL','AMT_CREDIT','AMT_ANNUITY','AMT_GOODS_PRICE']
sns.heatmap(pa_df[col1].corr(), annot=True,cmap="YlGn")

In [ ]:
#Heat map on numerical column for TARGET as 0

plt.figure(figsize=(15,12))
col1=['AMT_INCOME_TOTAL','AMT_CREDIT','AMT_ANNUITY','AMT_GOODS_PRICE']
sns.heatmap(zero_df[col1].corr(), annot=True,cmap="YlGn")

# PREVIOUS DATA


Read the dataset file previous_application.csv which consist previous loan of the customer.

In [ ]:
prev_df=pd.read_csv("./previous_application.csv")

prev_df.head()

In [ ]:
prev_df.shape

In [ ]:
freq_table4=prev_df.groupby(['NAME_CONTRACT_TYPE']).size().reset_index(name='count').rename(columns={'NAME_CONTRACT_TYPE':'CONTRACT_TYPE'})
freq_table4

##  Univariate Analysis

In [ ]:
plt.figure(figsize=(10,10))
sns.set(style="whitegrid")
ax=sns.barplot(x='CONTRACT_TYPE',y='count',data=freq_table4)
plt.title(['CONTRACT_TYPE'])

In [ ]:
freq_table4=prev_df.groupby(['CHANNEL_TYPE']).size().reset_index(name='count').rename(columns={'CHANNEL_TYPE':'CHANNEL_STATUS'})
freq_table4

In [ ]:
plt.figure(figsize=(15,10))
sns.set(style="whitegrid")
ax=sns.barplot(x='CHANNEL_STATUS',y='count',data=freq_table4)
plt.title(['CHANNEL_STATUS'])

## Bivariate Analysis

In [ ]:
plt.figure(figsize=(15,12))
col1=['AMT_APPLICATION','AMT_GOODS_PRICE','AMT_CREDIT','AMT_ANNUITY']
sns.heatmap(prev_df[col1].corr(), annot=True,cmap="YlGn")

In [ ]:
corr_columns=['AMT_APPLICATION','AMT_GOODS_PRICE','AMT_CREDIT','AMT_ANNUITY']

corr=prev_df[corr_columns].corr()
corr.style.background_gradient(cmap='coolwarm')

In [ ]:
sns.relplot(x="NAME_CONTRACT_TYPE", y="AMT_CREDIT", data=prev_df)

In [ ]:
sns.relplot(x='NAME_CONTRACT_STATUS', y="AMT_CREDIT",data=prev_df)

In [ ]:
sns.relplot(x='AMT_APPLICATION', y="AMT_CREDIT",data=prev_df)

In [ ]:
plt.figure(figsize=(15,10))

sns.boxplot(data=prev_df,x='WEEKDAY_APPR_PROCESS_START',y='AMT_CREDIT',hue='NAME_CONTRACT_TYPE')
plt.title('Difficulty on credit loan of different type')
plt.show()

# Conclusion


This case study aims to identify the defaulters and non-defaulters and develop a basic understanding of risk analytics in banking and financial services and understand how data is used to mini-mise the risk of losing money while lending to customers. After our  Analysis on the dataset we accept following key points:

1-Repeater customers are not likely to face any payment issues.                              
2-Customers who has family status as married having high chances of approved the loan.                    
3-Customers with age between 25-30 are likely to be not defaulter.                                 
4-Customer with low income(0-5 lakh) are those who applied for the loan.                                
5-Student and unemployed customers loan application is rejected most of the time and working customers loan will approved easily.                                                                                      
6-Most of the customers are used credit and cash medium in their loan.                                   
7-Thursday and Wednesday are most busiest day for the company so it is recommended that they increase their productivity on these weekdays.                                                                               
8-Most of the customer applied consumer and cash loan which share approx. equal number of approval and disapproval.    
9-Customers who provide half number of required documents are facing issue in their loan approval and customer who provide all the documents are get their approval immediately.      